GHzDACs/Cleanup/dac.py

```diff
@@ -936,8 +1024,534 @@ class DacRunner_Build11(DacRunner_Build7):
+            raise ValueError("From address must be {}<x<{}, ({} given)".format(
+                cls.JT_MIN_FROM_ADDR*4, cls.JT_MAX_FROM_ADDR*4, x_ns
+            ))
+        return x_ns // 4 + cls.JT_FROM_ADDR_OFFSET
+
+    @classmethod
+    def convert_to_address(cls, x_ns):
+        """ Convert a to address from ns to SRAM index.
+
+        This also checks bounds and applies offsets (there are currently no
+        offsets for to addresses).
+
+        :param int x_ns: to address in ns
+        :return: to address in SRAM units, corrected.
+        :rtype: int
+        """
+        if x_ns % 4 != 0:
+            raise ValueError("To address must be divisible by 4 ({} given)".format(x_ns))
+        if not(cls.JT_MIN_TO_ADDR <= x_ns // 4 <= cls.JT_MAX_TO_ADDR):
+            raise ValueError("To address must be {}<x<{}, ({} given)".format(
+                cls.JT_MIN_TO_ADDR*4, cls.JT_MAX_TO_ADDR*4, x_ns
+            ))
+        return x_ns // 4
+
+    @classmethod
+    def convert_end_address(cls, x_ns):
+        """ Convert a from address _for the END command_ from ns to SRAM index.
+
+        This also checks bounds and applies offsets.
+
+        :param int x_ns: from address (for the END command) in ns
+        :return: from address in SRAM units, corrected.
+        :rtype: int
+        """
+        if x_ns % 4 != 0:
+            raise ValueError("End address must be divisible by 4 ({} given)".format(x_ns))
+        if not(cls.JT_MIN_END_ADDR <= x_ns // 4 <= cls.JT_MAX_END_ADDR):
+            raise ValueError("End address must be {}<x<{}, ({} given)".format(
+                cls.JT_MIN_END_ADDR*4, cls.JT_MAX_END_ADDR*4, x_ns
+            ))
+        return x_ns // 4 + cls.JT_END_ADDR_OFFSET
+
+    @classmethod
+    def convert_idle_duration(cls, x_ns):
+        """ Convert an idle duration from ns to clock cycles.
+
+        This also checks bounds and applies offsets.
+
+        :param int x_ns: idle duration in ns
+        :return: idle duration in clock cycles, corrected
+        :rtype: int
+        """
+        if x_ns % 4 != 0:
+            raise ValueError("IDLE duration must be divisible by 4 ({} given)".format(x_ns))
+        if not(cls.JT_IDLE_MIN <= x_ns // 4 <= cls.JT_IDLE_MAX):
+            raise ValueError("Idle duration must be {}<x<{}, ({} given)".format(
+                cls.JT_IDLE_MIN*4, cls.JT_IDLE_MAX*4, x_ns
+            ))
+        return x_ns // 4 + cls.JT_IDLE_OFFSET
+
+    def buildRunner(self, reps, info):
+        jt_entries = info['jt_entries']
+        jt_counters = info['jt_counters']
+        start_delay = info.get('startDelay', 0)
+        loop_delay = info.get('loop_delay', 0)
+        sram = info.get('sram', None)
+        runner = self.RUNNER_CLASS(self, reps, start_delay, jt_entries, jt_counters, sram, loop_delay=loop_delay)
+        return runner
+
+    # let's check the various registry functions.
+    # OK means old implementation should work
```

Comment:
This was during development, when I was making a list to myself of "functions to check to see if they need to be updated for the JT register format". I will remove them now.